### PR TITLE
Share Caching

### PR DESCRIPTION
**Pull request by moshe-blox**
This PR rewrites the shares storage implementation to maintain a replica of the stored shares in-memory, skipping the database entirely in read operations.

### CODE REVIEW

File: operator/validator/controller.go [hunk 1]
```diff
@@ -266,19 +266,16 @@ func (c *controller) setupNetworkHandlers(logger *zap.Logger) error {
 	return nil
 }
 
-func (c *controller) GetOperatorShares(logger *zap.Logger) ([]*types.SSVShare, error) {
-	return c.sharesStorage.GetFilteredShares(logger, registrystorage.ByOperatorIDAndActive(c.operatorData.ID))
+func (c *controller) GetOperatorShares(logger *zap.Logger) []*types.SSVShare {
```
Comment (y0sher) [inline operator/validator/controller.go:269]: this method doesn't need `logger` anymore. (thank god.)

File: operator/validator/controller.go [hunk 2]
```diff
@@ -266,19 +266,16 @@ func (c *controller) setupNetworkHandlers(logger *zap.Logger) error {
...
+	return c.sharesStorage.List(registrystorage.ByOperatorID(c.operatorData.ID), registrystorage.ByActiveValidator())
 }
 
 func (c *controller) GetOperatorData() *registrystorage.OperatorData {
 	return c.operatorData
 }
 
 func (c *controller) GetValidatorStats(logger *zap.Logger) (uint64, uint64, uint64, error) {
```
Comment (y0sher) [inline operator/validator/controller.go:277]: also here `logger` not used or needed now.

File: registry/storage/shares.go [hunk 3]
```diff
@@ -8,130 +8,155 @@ import (
...
-// GetShare by key
-func (s *sharesStorage) GetShare(key []byte) (*types.SSVShare, bool, error) {
-	s.lock.RLock()
-	defer s.lock.RUnlock()
+func (s *sharesStorage) Get(pubKey []byte) *types.SSVShare {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 
-	return s.getUnsafe(key)
+	return s.shares[hex.EncodeToString(pubKey)]
```
Comment (y0sher) [inline registry/storage/shares.go:80]: so all the shares are always in the cache right?

Comment (moshe-blox) [reply registry/storage/shares.go:80]: > so all the shares are always in the cache right?

yeah, NewSharesStorage calls load which reads all shares from storage to memory, and each save updates both memory and storage

File: registry/storage/shares.go [hunk 4]
```diff
@@ -8,130 +8,155 @@ import (
...
+		match := true
+		for _, filter := range filters {
+			if !filter(share) {
+				match = false
+				break
+			}
+		}
+		if match {
+			shares = append(shares, share)
+		}
```
Comment (y0sher) [inline registry/storage/shares.go:102]:
```suggestion
        shareLoop:
	for _, share := range s.shares {
		for _, filter := range filters {
			if !filter(share) {
				continue shareLoop
			}
		}
		shares = append(shares, share)
```

**Review by y0sher — APPROVED** [review]

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -40,7 +40,7 @@ Build stage Docker image:
     - docker tag $IMAGE_NAME:$CI_BUILD_REF $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
     - $DOCKER_LOGIN_TO_INFRA_STAGE_REPO && docker push $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
   only:
-    - stage
+    - share-cache
```
Comment (nkryuchkov) [inline .gitlab-ci.yml:43]: @moshe-blox not reverted

**Pull request merged**